googleapiclient/http.py

```diff
@@ -82,22 +89,65 @@ def _retry_request(http, num_retries, req_type, sleep, rand, uri, method, *args,
     resp, content - Response from the http request (may be HTTP 5xx).
   """
   resp = None
+  content = None
   for retry_num in range(num_retries + 1):
     if retry_num > 0:
-      sleep(rand() * 2**retry_num)
+      # Sleep before retrying.
+      sleep_time = rand() * 2 ** retry_num
       logging.warning(
-          'Retry #%d for %s: %s %s%s' % (retry_num, req_type, method, uri,
+          'Sleeping %.2f seconds before retry %d of %d for %s: %s %s%s' %
+          (sleep_time, retry_num, num_retries, req_type, method, uri,
           ', following status: %d' % resp.status if resp else ''))
+      sleep(sleep_time)
 
     try:
       resp, content = http.request(uri, method, *args, **kwargs)
-    except ssl.SSLError:
+    # Retry on SSL errors and socket timeout errors.
+    except (ssl.SSLError, TimeoutError) as ex:
+      # If this is a socket.error, only retry if it's a timeout.
+      if isinstance(ex, socket.error) and not isinstance(ex, ssl.SSLError):
+        try:
+          # errno's contents differ by platform so we have to match by name.
+          if not (socket.errno.errorcode[ex.errno] in
+              ['WSAETIMEDOUT', 'ETIMEDOUT']):
+            raise
+        # If we couldn't match it - retry because we can't know for sure.
+        except KeyError:
+          pass
       if retry_num == num_retries:
         raise
       else:
+        logging.warning('Encountered exception %s', repr(ex))
         continue
-    if resp.status < 500:
-      break
+
+    # Retry 5xx errors.
+    if resp.status >= 500:
+      continue
+
+    # Retry 429 errors.
+    if resp.status == 429:
+      continue
+
+    # For 403 errors, we have to check for the `reason` in the response to
+    # determine if it is safe to retry.
+    if resp.status == 403:
+      # If there's no details about the 403 type, don't retry.
+      if not content:
+        break
+
+      # Content is in JSON format.
+      try:
+        data = json.loads(content.decode('utf-8'))
+        reason = data['error']['errors'][0]['reason']
+      except (UnicodeDecodeError, ValueError, KeyError):
+        logging.warning('Invalid JSON content from response')
+        break
+
+      logging.warning('Encountered status 403 with reason "%s"' % reason)

```

Comment:
`logging.warning(<format string>, <arguments to format string>)` rather than `logging.warning(<format string> % <arguments to format string>)`.

Fun fact: long ago in a different life I was the one who wrote the patch to pylint that inspects code for this issue.